shpkpr/marathon/application.py

```diff
@@ -0,0 +1,74 @@
+# stdlib imports
+import json
+import os
+
+# third-party imports
+import jsonschema
+from cached_property import cached_property
+
+# local imports
+from shpkpr import exceptions
+
+
+def default_schema():
+    """Path to the default schema used to validate application definitions
+    before sending to marathon.
+    """
+    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "schema", "app.json")
+
+
+class ValidationError(exceptions.ShpkprException):
+    exit_code = 2
+
+    def format_message(self):
+        return 'Unable to validate application: %s' % self.message
+
+
+class ApplicationError(exceptions.ShpkprException):
+    exit_code = 2
+
+
+class MarathonApplication(object):
```

Comment:
Been thinking about this one, and the more I do, the more I like the idea of just using a plain old python dictionary to store and pass around the application data.

It's how the data comes in (whether from a remote API call, or a rendered template), and how the data leaves when we post it off to marathon. Transforming that data into concrete classes just so we can transform it back again almost immediately feels like asking for trouble (the best code is no code), I don't think we'd gain much by doing so.

Using a plain dictionary in combination with strict JSON schema validation at the boundaries of the application seems like a good combination of convenience and safety to me. We're not a million miles away from that right now, just need those extra few steps to make things safer.

The only issue I see with that approach is that our current JSON schema only covers the fields that can be set when deploying an application, this would need to be expanded to cover all fields returned by the marathon API so we can validate incoming data too.